native/membrane_videocompositor_wgpu/src/lib.rs

```diff
@@ -62,41 +58,116 @@ impl InnerState {
-        rustler::ResourceArc::new(State::new(upper_caps, lower_caps, output_caps)),
+        rustler::ResourceArc::new(State::new(output_caps)),
     ))
 }
 
 #[rustler::nif]
 fn join_frames<'a>(
     env: rustler::Env<'a>,
     state: ResourceArc<State>,
-    upper: rustler::Binary,
-    lower: rustler::Binary,
+    frames: Vec<(usize, rustler::Binary)>,
 ) -> Result<(rustler::Atom, rustler::Term<'a>), rustler::Error> {
     let state = state.lock().unwrap();
 
+    for (i, frame) in &frames {
+        state.compositor.upload_texture(*i, frame)?
+    }
+
     let mut output =
         rustler::OwnedBinary::new(state.output_caps.width * state.output_caps.height * 3 / 2)
             .unwrap(); //FIXME: return an error instead of panicking here
 
-    pollster::block_on(state.compositor.join_frames(
-        upper.as_slice(),
-        lower.as_slice(),
-        output.as_mut_slice(),
-    ));
+    pollster::block_on(state.compositor.draw_into(output.as_mut_slice()));
 
     Ok((atoms::ok(), output.release(env).to_term(env)))
 }
 
+#[rustler::nif]
+fn add_video(
+    #[allow(unused)] env: rustler::Env<'_>,
+    state: rustler::ResourceArc<State>,
+    id: usize,
+    input_video: RawVideo,
+    position: Position,
+) -> Result<rustler::Atom, rustler::Error> {
+    let mut state: std::sync::MutexGuard<InnerState> = state.lock().unwrap();
+    let placement = determine_video_placement(&state, &input_video, &position);
+
+    state
+        .compositor
+        .add_video(id, placement, input_video.width, input_video.height);
+    Ok(atoms::ok())
+}
+
+/// Maps point `x` from the domain \[`x_min`, `x_max`\] to the point in the \[`y_min, y_max`\] line segment, using linear interpolation.
+///
+/// `x` outside the original domain will be extrapolated outside the targe domain.
+fn lerp(x: f64, x_min: f64, x_max: f64, y_min: f64, y_max: f64) -> f64 {
+    (x - x_min) / (x_max - x_min) * (y_max - y_min) + y_min
+}
+
+fn determine_video_placement(
+    state: &InnerState,
+    input_video: &RawVideo,
+    position: &Position,
+) -> compositor::VideoPlacementTemplate {
+    let scene_width = state.output_caps.width;
+    let scene_height = state.output_caps.height;
+
+    let left = lerp(position.x as f64, 0.0, scene_width as f64, -1.0, 1.0) as f32;
+    let right = lerp(
+        (position.x + input_video.width) as f64,
+        0.0,
+        scene_width as f64,
+        -1.0,
+        1.0,
+    ) as f32;
+    let top = lerp(position.y as f64, 0.0, scene_height as f64, 1.0, -1.0) as f32;
+    let bot = lerp(
+        (position.y + input_video.height) as f64,
+        0.0,
+        scene_height as f64,
+        1.0,
+        -1.0,
+    ) as f32;
+
+    compositor::VideoPlacementTemplate {
+        top_right: compositor::Point(right, top),
+        top_left: compositor::Point(left, top),
+        bot_left: compositor::Point(left, bot),
+        bot_right: compositor::Point(right, bot),
+        z_value: 0.0,
+    }
+}
+
+#[rustler::nif]
+fn set_position(
```

Comment:
Can't we just update video.vertices instead of returning error?

Reply:
We could, but it wouldn't be trivial to implement, and this isn't used anywhere, and I think even @Janix4000 agrees that adding this to the `FrameCompositor` behaviour was done a bit too hastily and this isn't very useful or necessary.

Reply:
I think that changing video position is one of the key features of a video compositor. As a said on today's daily, I think that such functions should be implemented in `State impl` (maybe with some call to `InputVideo impl`), rather than in `lib.ex` (though I think we need Dominik's opinion on that). We need that feature even if elixir behaviour changes and by implementing it in that way, we'll need to change only integration with elixir.

Reply:
I honestly think this is a feature we'll never use in this form and will need to rewrite completely when switching to the new architecture.